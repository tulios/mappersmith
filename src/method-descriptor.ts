import type { Headers, RequestParams, ParameterEncoderFn, Params } from './types'
import type { Middleware } from './middleware/index'

export interface MethodDescriptorParams {
  allowResourceHostOverride?: boolean
  authAttr?: string
  binary?: boolean
  bodyAttr?: string
  headers?: Headers
  headersAttr?: string
  host: string
  hostAttr?: string
  method?: string
  middleware?: Array<Middleware>
  middlewares?: Array<Middleware>
  parameterEncoder?: ParameterEncoderFn
  params?: Params
  path: string | ((args: RequestParams) => string)
  pathAttr?: string
  queryParamAlias?: Record<string, string>
  signalAttr?: string
  timeoutAttr?: string
}

/**
 * @typedef MethodDescriptor
 * @param {MethodDescriptorParams} params
 *   @param {boolean} params.allowResourceHostOverride
 *   @param {Function} params.parameterEncoder
 *   @param {String} params.authAttr - auth attribute name. Default: 'auth'
 *   @param {boolean} params.binary
 *   @param {String} params.bodyAttr - body attribute name. Default: 'body'
 *   @param {Headers} params.headers
 *   @param {String} params.headersAttr - headers attribute name. Default: 'headers'
 *   @param {String} params.host
 *   @param {String} params.hostAttr - host attribute name. Default: 'host'
 *   @param {String} params.method
 *   @param {Middleware[]} params.middleware
 *   @param {Middleware[]} params.middlewares - alias for middleware
 *   @param {RequestParams} params.params
 *   @param {String|Function} params.path
 *   @param {String} params.pathAttr. Default: 'path'
 *   @param {Object} params.queryParamAlias
 *   @param {Number} params.signalAttr - signal attribute name. Default: 'signal'
 *   @param {Number} params.timeoutAttr - timeout attribute name. Default: 'timeout'
 */
export class MethodDescriptor {
  public readonly allowResourceHostOverride: boolean
  public readonly authAttr: string
  public readonly binary: boolean
  public readonly bodyAttr: string
  public readonly headers?: Headers
  public readonly headersAttr: string
  public readonly host: string
  public readonly hostAttr: string
  public readonly method: string
  public readonly middleware: Middleware[]
  public readonly parameterEncoder: ParameterEncoderFn
  public readonly params?: RequestParams
  public readonly path: string | ((args: RequestParams) => string)
  public readonly pathAttr: string
  public readonly queryParamAlias: Record<string, string>
  public readonly signalAttr: string
  public readonly timeoutAttr: string

  constructor(params: MethodDescriptorParams) {
    this.allowResourceHostOverride = params.allowResourceHostOverride || false
    this.binary = params.binary || false
    this.headers = params.headers
    this.host = params.host
    this.method = params.method || 'get'
    this.parameterEncoder = params.parameterEncoder || encodeURIComponent
    this.params = params.params
    this.path = params.path
    this.queryParamAlias = params.queryParamAlias || {}

    this.authAttr = params.authAttr || 'auth'
    this.bodyAttr = params.bodyAttr || 'body'
    this.headersAttr = params.headersAttr || 'headers'
    this.hostAttr = params.hostAttr || 'host'
    this.pathAttr = params.pathAttr || 'path'
    this.signalAttr = params.signalAttr || 'signal'
    this.timeoutAttr = params.timeoutAttr || 'timeout'

    const resourceMiddleware = params.middleware || params.middlewares || []
    this.middleware = resourceMiddleware
  }
}

export default MethodDescriptor
