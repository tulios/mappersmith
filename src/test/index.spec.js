import { lookupResponse, mockClient, clear, install, m } from './index'
import { MethodDescriptor } from '../method-descriptor'
import { Request } from '../request'
import randomJsonData from './data/random-json-data.json'
import { forge } from '../index'

const stripAnsi = (str) => {
  // Regular expression to match ANSI escape codes
  // This removes all the color codes and other special codes
  // eslint-disable-next-line no-control-regex
  const ansiRegex = /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g
  return str.replace(ansiRegex, '')
}

const LOG_TO_CONSOLE = false

const expectToThrowErrorMatchingSnapshotWithoutAnsi = (fn, logToConsole = LOG_TO_CONSOLE) => {
  try {
    fn()
    throw new Error('Expected function to throw an error, but it did not.')
  } catch (error) {
    if (logToConsole) console.log(error)
    const cleanedError = stripAnsi(error.message)
    expect(cleanedError).toMatchSnapshot()
  }
}

describe('mappersmith/test/lookupResponse', () => {
  beforeEach(() => {
    clear()
    // Set this to a fixed value so that the tables are always the same width
    process.stdout.columns = 100
    install({ enhancedDebugging: true })
  })

  it('should match when request is exact match', () => {
    const method = 'GET'
    const host = 'http://example.org'
    const path = '/path'
    const body = { id: 1 }
    const responseBody = { id: 'my-id' }
    const headers = { 'X-TEST': 'VALUE' }
    const query = { id: 1 }

    mockClient(forge({ host, resources: { users: { get: { path, method } } } }))
      .resource('users')
      .method('get')
      .with({ body, headers, query })
      .response(responseBody)
      .status(200)

    const request = new Request(new MethodDescriptor({ method, host, path }), {
      body,
      headers,
      query,
    })

    expect(() => lookupResponse(request)).not.toThrow()

    const response = lookupResponse(request)
    expect(response.responseData).toEqual(JSON.stringify(responseBody))
  })

  it('should output a clear message when there are no mocks installed', () => {
    const method = 'GET'
    const host = 'http://example.org'
    const path = '/path'
    const body = {
      parameter: 1,
      deep: {
        maybe_gql: 'query { user { id } }',
        variables: { id: 1 },
      },
    }
    const headers = { header1: 'value', header2: 'value2' }
    const query = { parameter: 1, parameter2: 2 }

    const request = new Request(new MethodDescriptor({ method, host, path }), {
      body,
      headers,
      query,
    })

    expectToThrowErrorMatchingSnapshotWithoutAnsi(() => lookupResponse(request))
  })

  it('should log all existing mocks when there is no match', () => {
    const body = {
      parameter: 1,
      deep: {
        maybe_gql: 'query { user { id } }',
        variables: { id: 1 },
      },
    }
    const headers = { header1: 'value', header2: 'value2' }
    const query = { parameter: 1, parameter2: 2 }

    mockClient(
      forge({
        host: 'http://example-1.org',
        resources: { users: { get: { path: '/no-path', method: 'POST' } } },
      })
    )
      .resource('users')
      .method('get')
      .with({ body, headers, query })

    const request = new Request(
      new MethodDescriptor({ host: 'http://example.org', method: 'GET', path: '/path' }),
      { body, headers, query }
    )

    expectToThrowErrorMatchingSnapshotWithoutAnsi(() => lookupResponse(request))
  })

  it('should log all existing mocks when there is a partial match', () => {
    mockClient(
      forge({
        host: 'http://example.org',
        resources: {
          users: { get: { path: '/path', method: 'GET' } },
        },
      })
    )
      .resource('users')
      .method('get')
      .with({ body: { id: 1 }, headers: { 'X-TEST': 'VALUE' } })

    const request = new Request(
      new MethodDescriptor({ host: 'http://example.org', method: 'GET', path: '/path' }),
      {
        body: { id: 2 },
        headers: { 'X-TEST': 'VALUE2' },
      }
    )

    expectToThrowErrorMatchingSnapshotWithoutAnsi(() => lookupResponse(request))
  })

  it('should correctly handle m.anything() displaying that it matches ANYTHING', () => {
    mockClient(
      forge({
        host: 'http://example.org',
        resources: {
          users: { get: { path: '/path', method: 'GET' } },
        },
      })
    )
      .resource('users')
      .method('get')
      .with({ body: m.anything(), headers: { 'X-TEST': 'VALUE-2' } })

    const request = new Request(
      new MethodDescriptor({ host: 'http://example.org', method: 'GET', path: '/path' }),
      {
        body: { id: 2 },
        headers: { 'X-TEST': 'VALUE2' },
      }
    )

    expectToThrowErrorMatchingSnapshotWithoutAnsi(() => lookupResponse(request))
  })

  it('should handle multiple mocks with large bodies and long URLs, including a partial match', () => {
    const client = forge({
      clientId: 'My Client',
      host: 'http://example.org',
      resources: {
        users: { get: { path: '/path', method: 'GET' } },
        posts: { post: { path: '/long/path/to/resource/with/many/segments', method: 'POST' } },
        comments: { put: { path: '/another/very/long/path/to/resource', method: 'PUT' } },
      },
    })

    mockClient(client)
      .resource('users')
      .method('get')
      .with({
        body: { id: 1 },
        headers: { 'X-TEST': 'VALUE' },
        query: { id: 1 },
      })
      .response({ id: 'user-id' })
      .status(200)

    mockClient(client)
      .resource('posts')
      .method('post')
      .with({
        body: { content: 'This is a very large body with lots of content...' },
        headers: { 'Content-Type': 'application/json' },
        query: { postId: 12345 },
      })
      .response({ id: 'post-id' })
      .status(201)

    // Partial match: same URL and method, but different body and headers
    mockClient(client)
      .named('My Named Mock')
      .resource('users')
      .method('get')
      .with({
        body: randomJsonData,
        headers: { 'X-TEST': 'DIFFERENT-VALUE' },
      })
      .response({ id: 'partial-match-id' })
      .status(200)

    mockClient(client)
      .resource('comments')
      .method('put')
      .with({
        body: { ...randomJsonData, _id: '66ec092d7189db5c15ea3f05' },
        headers: { 'Content-Type': 'application/json' },
        query: { commentId: 67890 },
      })
      .response({ id: 'comment-id' })
      .status(200)

    const request = new Request(
      new MethodDescriptor({ host: 'http://example.org', method: 'GET', path: '/path' }),
      {
        body: randomJsonData,
        headers: { 'X-TEST': 'DIFFERENT-VALUE' },
        query: { id: 2 },
      }
    )

    expectToThrowErrorMatchingSnapshotWithoutAnsi(() => lookupResponse(request))
  })
})
