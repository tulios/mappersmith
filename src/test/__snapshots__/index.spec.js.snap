// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mappersmith/test/lookupResponse should correctly handle m.anything() displaying that it matches ANYTHING 1`] = `
"


------------------------------------------------------------------------------------------------------------------
Mappersmith matches a mock to an outgoing request by comparing the request's URL, METHOD, BODY and HEADERS.

URL:     The URL is sorted and normalized before comparison.
BODY:    The BODY is sorted and normalized in the form of a query-string before comparison.
HEADERS: The headers of the outgoing request is stripped of any headers that are not present in the 
         mock definition headers and then sorted and normalized before comparison.
METHOD:  The method of the outgoing request is compared as is.
------------------------------------------------------------------------------------------------------------------

Request:

  ┌────────────────────┬──────────────────────────────────────────────────────────────────────┐
  │  Outbound Request  │                                Value                                 │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Url                │ http://example.org/path                                              │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Method             │ get                                                                  │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Body               │ id=2                                                                 │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Headers            │ x-test=VALUE2                                                        │
  └────────────────────┴──────────────────────────────────────────────────────────────────────┘

Mock definitions installed:


  ┌───────────────────────────────────────────────┬───────────────────────────────────────────────┐
  │              Colors Explanation               │               Diff Explanation                │
  ├───────────────────────────────────────────────┼───────────────────────────────────────────────┤
  │                                               │                                               │
  │ ● Exact match                                 │ Mock value matches the request value          │
  │ ● Partial match (When URL and METHOD match)   │ Present in mock but not in request            │
  │ ● Not matching                                │ Present in request but not in mock            │
  │                                               │                                               │
  └───────────────────────────────────────────────┴───────────────────────────────────────────────┘


  ┌────────────────┬────────────┬─────────────────────────────────────────────────────────────────┐
  │   users - 4    │   Match    │                           Value/Diff                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Url            │ Yes        │ http://example.org/path                                         │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Method         │ Yes        │ get                                                             │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Body           │ Yes        │ MATCHED_BY_FUNCTION                                             │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Headers        │ No         │ x-test=VALUE-2                                                  │
  └────────────────┴────────────┴─────────────────────────────────────────────────────────────────┘


[Mappersmith Test] No exact match found but a partial match was found, check your mock definitions, debug information available above.
"
`;

exports[`mappersmith/test/lookupResponse should handle multiple mocks with large bodies and long URLs, including a partial match 1`] = `
"


------------------------------------------------------------------------------------------------------------------
Mappersmith matches a mock to an outgoing request by comparing the request's URL, METHOD, BODY and HEADERS.

URL:     The URL is sorted and normalized before comparison.
BODY:    The BODY is sorted and normalized in the form of a query-string before comparison.
HEADERS: The headers of the outgoing request is stripped of any headers that are not present in the 
         mock definition headers and then sorted and normalized before comparison.
METHOD:  The method of the outgoing request is compared as is.
------------------------------------------------------------------------------------------------------------------

Request:

  ┌────────────────────┬──────────────────────────────────────────────────────────────────────┐
  │  Outbound Request  │                                Value                                 │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Url                │ http://example.org/path?query[id]=2                                  │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Method             │ get                                                                  │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Body               │ _id=66ec092d7189db5c15ea3f04&about=Cupidatat+excepteur+do+excepteur+ │
  │                    │ quis+deserunt+nisi+aliqua+nulla+ut+esse.+Non+ea+tempor+occaecat+sunt │
  │                    │ +commodo.+Nisi+cillum+non+anim+aute+laborum+pariatur+irure+occaecat+ │
  │                    │ minim.                                                               │
  │                    │ &address=377+Conduit+Boulevard,+Franklin,+Alabama,+7940&age=31&balan │
  │                    │ ce=$1,349.22&company=WATERBABY&email=deborahcantu@waterbaby.com&eyeC │
  │                    │ olor=green&favoriteFruit=banana&friends[][id]=0&friends[][name]=Marq │
  │                    │ uez+Langley&friends[][id]=1&friends[][name]=Blackburn+Oconnor&friend │
  │                    │ s[][id]=2&friends[][name]=Tameka+Mcdonald&gender=female&greeting=Hel │
  │                    │ lo,+Deborah+Cantu!+You+have+1+unread+messages.&guid=3f6091f3-a629-46 │
  │                    │ 53-913f-eb797bede6e9&index=6&isActive=true&latitude=26.758942&longit │
  │                    │ ude=29.872733&name=Deborah+Cantu&phone=+1+(951)+590-3102&picture=htt │
  │                    │ p://placehold.it/32x32&registered=2014-10-31T02:57:12+-01:00&tags[]= │
  │                    │ pariatur&tags[]=Lorem&tags[]=tempor&tags[]=ut&tags[]=veniam&tags[]=t │
  │                    │ empor&tags[]=amet                                                    │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Headers            │ x-test=DIFFERENT-VALUE                                               │
  └────────────────────┴──────────────────────────────────────────────────────────────────────┘

Mock definitions installed:


  ┌───────────────────────────────────────────────┬───────────────────────────────────────────────┐
  │              Colors Explanation               │               Diff Explanation                │
  ├───────────────────────────────────────────────┼───────────────────────────────────────────────┤
  │                                               │                                               │
  │ ● Exact match                                 │ Mock value matches the request value          │
  │ ● Partial match (When URL and METHOD match)   │ Present in mock but not in request            │
  │ ● Not matching                                │ Present in request but not in mock            │
  │                                               │                                               │
  └───────────────────────────────────────────────┴───────────────────────────────────────────────┘


  ┌────────────────┬────────────┬─────────────────────────────────────────────────────────────────┐
  │  My Client -   │   Match    │                           Value/Diff                            │
  │     posts      │            │                                                                 │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Url            │ No         │ http://example.org/long/path/to/resource/with/many/segments?que │
  │                │            │ ry[ipostId]=12345                                               │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Method         │ No         │ gepost                                                          │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Body           │ No         │ _id=66ec092d7189db5c15ea3f04&abount=Cupidatat+excepteur+do+exce │
  │                │            │ pteur+quis+deserunt+n=Thisi+aliqua+nulla+ut+esse.+Non+ea+tvempo │
  │                │            │ ry+occaecat+sunt+commodo.+Nisi+cillum+non+anim+autrge+laborum+p │
  │                │            │ ariatur+irure+occaecat+minim.                                   │
  │                │            │ &address=377y+Conduwith+Boulevard,+Franklin,+Alabama,+7940&age= │
  │                │            │ 31&balance=$1,349.22&company=WATERBABY&email=deborahcantu@water │
  │                │            │ baby.com&eyeColor=green&favoriteFruit=banana&friends[][id]=0&fr │
  │                │            │ iends[][name]=Marquez+Langley&friends[][id]=1&friends[][name]=B │
  │                │            │ lackburn+Oconnor&friends[][id]=2&friends[][name]=Tameka+Mcdonal │
  │                │            │ d&gender=female&greeting=Hello,+Deborah+Cantu!+You+have+1+unrea │
  │                │            │ d+messages.&guid=3f6091f3-a629-4653-913f-eb797bede6e9&index=6&i │
  │                │            │ sActive=true&latitude=26.758942&longitude=29.872733&name=Debora │
  │                │            │ h+Cantu&phone=+1+(951)+590-3102&picture=http://placehold.it/32x │
  │                │            │ 32&registered=2014-10-31T02:57:12+-01:00&tags[]=pariatur&tags[] │
  │                │            │ =Lorem&tags[]=tempor&tags[]=ut&tags[]=veniam&tags[]=tempor&tags │
  │                │            │ []=amet                                                         │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Headers        │ No         │ content-type=application%2Fjson                                 │
  └────────────────┴────────────┴─────────────────────────────────────────────────────────────────┘

  ┌────────────────┬────────────┬─────────────────────────────────────────────────────────────────┐
  │  My Client -   │   Match    │                           Value/Diff                            │
  │    comments    │            │                                                                 │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Url            │ No         │ http://example.org/another/very/long/path/to/resource?query[ico │
  │                │            │ mmentId]=267890                                                 │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Method         │ No         │ geput                                                           │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Body           │ No         │ _id=66ec092d7189db5c15ea3f045&about=Cupidatat+excepteur+do+exce │
  │                │            │ pteur+quis+deserunt+nisi+aliqua+nulla+ut+esse.+Non+ea+tempor+oc │
  │                │            │ caecat+sunt+commodo.+Nisi+cillum+non+anim+aute+laborum+pariatur │
  │                │            │ +irure+occaecat+minim.                                          │
  │                │            │ &address=377+Conduit+Boulevard,+Franklin,+Alabama,+7940&age=31& │
  │                │            │ balance=$1,349.22&company=WATERBABY&email=deborahcantu@waterbab │
  │                │            │ y.com&eyeColor=green&favoriteFruit=banana&friends[][id]=0&frien │
  │                │            │ ds[][name]=Marquez+Langley&friends[][id]=1&friends[][name]=Blac │
  │                │            │ kburn+Oconnor&friends[][id]=2&friends[][name]=Tameka+Mcdonald&g │
  │                │            │ ender=female&greeting=Hello,+Deborah+Cantu!+You+have+1+unread+m │
  │                │            │ essages.&guid=3f6091f3-a629-4653-913f-eb797bede6e9&index=6&isAc │
  │                │            │ tive=true&latitude=26.758942&longitude=29.872733&name=Deborah+C │
  │                │            │ antu&phone=+1+(951)+590-3102&picture=http://placehold.it/32x32& │
  │                │            │ registered=2014-10-31T02:57:12+-01:00&tags[]=pariatur&tags[]=Lo │
  │                │            │ rem&tags[]=tempor&tags[]=ut&tags[]=veniam&tags[]=tempor&tags[]= │
  │                │            │ amet                                                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Headers        │ No         │ content-type=application%2Fjson                                 │
  └────────────────┴────────────┴─────────────────────────────────────────────────────────────────┘

  ┌────────────────┬────────────┬─────────────────────────────────────────────────────────────────┐
  │  My Client -   │   Match    │                           Value/Diff                            │
  │     users      │            │                                                                 │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Url            │ No         │ http://example.org/path?query[id]=21                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Method         │ Yes        │ get                                                             │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Body           │ No         │ _id=66ec092d7189db5c15ea3f04&about=Cupidatat+excepteur+do+excep │
  │                │            │ teur+quis+deserunt+nisi+aliqua+nulla+ut+esse.+Non+ea+tempor+occ │
  │                │            │ aecat+sunt+commodo.+Nisi+cillum+non+anim+aute+laborum+pariatur+ │
  │                │            │ irure+occaecat+minim.                                           │
  │                │            │ &address=377+Conduit+Boulevard,+Franklin,+Alabama,+7940&age=31& │
  │                │            │ balance=$1,349.22&company=WATERBABY&email=deborahcantu@waterbab │
  │                │            │ y.com&eyeColor=green&favoriteFruit=banana&friends[][id]=0&frien │
  │                │            │ ds[][name]=Marquez+Langley&friends[][id]=1&friends[][name]=Blac │
  │                │            │ kburn+Oconnor&friends[][id]=2&friends[][name]=Tameka+Mcdonald&g │
  │                │            │ ender=female&greeting=Hello,+Deborah+Cantu!+You+have+1+unread+m │
  │                │            │ essages.&guid=3f6091f3-a629-4653-913f-eb797bede6e9&index=6&isAc │
  │                │            │ tive=true&latitude=26.758942&longitude=29.872733&name=Deborah+C │
  │                │            │ antu&phone=+1+(951)+590-3102&picture=http://placehold.it/32x32& │
  │                │            │ registered=2014-10-31T02:57:12+-01:00&tags[]=pariatur&tags[]=Lo │
  │                │            │ rem&tags[]=tempor&tags[]=ut&tags[]=veniam&tags[]=tempor&tags[]= │
  │                │            │ amet                                                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Headers        │ No         │ x-test=DIFFERENT-VALUE                                          │
  └────────────────┴────────────┴─────────────────────────────────────────────────────────────────┘

  ┌────────────────┬────────────┬─────────────────────────────────────────────────────────────────┐
  │ My Named Mock  │   Match    │                           Value/Diff                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Url            │ No         │ http://example.org/path?query[id]=2                             │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Method         │ Yes        │ get                                                             │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Body           │ Yes        │ _id=66ec092d7189db5c15ea3f04&about=Cupidatat+excepteur+do+excep │
  │                │            │ teur+quis+deserunt+nisi+aliqua+nulla+ut+esse.+Non+ea+tempor+occ │
  │                │            │ aecat+sunt+commodo.+Nisi+cillum+non+anim+aute+laborum+pariatur+ │
  │                │            │ irure+occaecat+minim.                                           │
  │                │            │ &address=377+Conduit+Boulevard,+Franklin,+Alabama,+7940&age=31& │
  │                │            │ balance=$1,349.22&company=WATERBABY&email=deborahcantu@waterbab │
  │                │            │ y.com&eyeColor=green&favoriteFruit=banana&friends[][id]=0&frien │
  │                │            │ ds[][name]=Marquez+Langley&friends[][id]=1&friends[][name]=Blac │
  │                │            │ kburn+Oconnor&friends[][id]=2&friends[][name]=Tameka+Mcdonald&g │
  │                │            │ ender=female&greeting=Hello,+Deborah+Cantu!+You+have+1+unread+m │
  │                │            │ essages.&guid=3f6091f3-a629-4653-913f-eb797bede6e9&index=6&isAc │
  │                │            │ tive=true&latitude=26.758942&longitude=29.872733&name=Deborah+C │
  │                │            │ antu&phone=+1+(951)+590-3102&picture=http://placehold.it/32x32& │
  │                │            │ registered=2014-10-31T02:57:12+-01:00&tags[]=pariatur&tags[]=Lo │
  │                │            │ rem&tags[]=tempor&tags[]=ut&tags[]=veniam&tags[]=tempor&tags[]= │
  │                │            │ amet                                                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Headers        │ Yes        │ x-test=DIFFERENT-VALUE                                          │
  └────────────────┴────────────┴─────────────────────────────────────────────────────────────────┘


[Mappersmith Test] No exact match found but a partial match was found, check your mock definitions, debug information available above.
"
`;

exports[`mappersmith/test/lookupResponse should log all existing mocks when there is a partial match 1`] = `
"


------------------------------------------------------------------------------------------------------------------
Mappersmith matches a mock to an outgoing request by comparing the request's URL, METHOD, BODY and HEADERS.

URL:     The URL is sorted and normalized before comparison.
BODY:    The BODY is sorted and normalized in the form of a query-string before comparison.
HEADERS: The headers of the outgoing request is stripped of any headers that are not present in the 
         mock definition headers and then sorted and normalized before comparison.
METHOD:  The method of the outgoing request is compared as is.
------------------------------------------------------------------------------------------------------------------

Request:

  ┌────────────────────┬──────────────────────────────────────────────────────────────────────┐
  │  Outbound Request  │                                Value                                 │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Url                │ http://example.org/path                                              │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Method             │ get                                                                  │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Body               │ id=2                                                                 │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Headers            │ x-test=VALUE2                                                        │
  └────────────────────┴──────────────────────────────────────────────────────────────────────┘

Mock definitions installed:


  ┌───────────────────────────────────────────────┬───────────────────────────────────────────────┐
  │              Colors Explanation               │               Diff Explanation                │
  ├───────────────────────────────────────────────┼───────────────────────────────────────────────┤
  │                                               │                                               │
  │ ● Exact match                                 │ Mock value matches the request value          │
  │ ● Partial match (When URL and METHOD match)   │ Present in mock but not in request            │
  │ ● Not matching                                │ Present in request but not in mock            │
  │                                               │                                               │
  └───────────────────────────────────────────────┴───────────────────────────────────────────────┘


  ┌────────────────┬────────────┬─────────────────────────────────────────────────────────────────┐
  │   users - 3    │   Match    │                           Value/Diff                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Url            │ Yes        │ http://example.org/path                                         │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Method         │ Yes        │ get                                                             │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Body           │ No         │ id=21                                                           │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Headers        │ No         │ x-test=VALUE2                                                   │
  └────────────────┴────────────┴─────────────────────────────────────────────────────────────────┘


[Mappersmith Test] No exact match found but a partial match was found, check your mock definitions, debug information available above.
"
`;

exports[`mappersmith/test/lookupResponse should log all existing mocks when there is no match 1`] = `
"


------------------------------------------------------------------------------------------------------------------
Mappersmith matches a mock to an outgoing request by comparing the request's URL, METHOD, BODY and HEADERS.

URL:     The URL is sorted and normalized before comparison.
BODY:    The BODY is sorted and normalized in the form of a query-string before comparison.
HEADERS: The headers of the outgoing request is stripped of any headers that are not present in the 
         mock definition headers and then sorted and normalized before comparison.
METHOD:  The method of the outgoing request is compared as is.
------------------------------------------------------------------------------------------------------------------

Request:

  ┌────────────────────┬──────────────────────────────────────────────────────────────────────┐
  │  Outbound Request  │                                Value                                 │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Url                │ http://example.org/path?query[parameter]=1&query[parameter2]=2       │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Method             │ get                                                                  │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Body               │ maybe_gql=query+{+user+{+id+}+}&id=1&parameter=1                     │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Headers            │ header1=value&header2=value2                                         │
  └────────────────────┴──────────────────────────────────────────────────────────────────────┘

Mock definitions installed:


  ┌───────────────────────────────────────────────┬───────────────────────────────────────────────┐
  │              Colors Explanation               │               Diff Explanation                │
  ├───────────────────────────────────────────────┼───────────────────────────────────────────────┤
  │                                               │                                               │
  │ ● Exact match                                 │ Mock value matches the request value          │
  │ ● Partial match (When URL and METHOD match)   │ Present in mock but not in request            │
  │ ● Not matching                                │ Present in request but not in mock            │
  │                                               │                                               │
  └───────────────────────────────────────────────┴───────────────────────────────────────────────┘


  ┌────────────────┬────────────┬─────────────────────────────────────────────────────────────────┐
  │   users - 2    │   Match    │                           Value/Diff                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Url            │ No         │ http://example-1.org/no-path?query[parameter]=1&query[parameter │
  │                │            │ 2]=2                                                            │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Method         │ No         │ gepost                                                          │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Body           │ Yes        │ maybe_gql=query+{+user+{+id+}+}&id=1&parameter=1                │
  ├────────────────┼────────────┼─────────────────────────────────────────────────────────────────┤
  │ Headers        │ Yes        │ header1=value&header2=value2                                    │
  └────────────────┴────────────┴─────────────────────────────────────────────────────────────────┘


[Mappersmith Test] No match found, check your mock definitions, debug information available above.
"
`;

exports[`mappersmith/test/lookupResponse should output a clear message when there are no mocks installed 1`] = `
"


------------------------------------------------------------------------------------------------------------------
Mappersmith matches a mock to an outgoing request by comparing the request's URL, METHOD, BODY and HEADERS.

URL:     The URL is sorted and normalized before comparison.
BODY:    The BODY is sorted and normalized in the form of a query-string before comparison.
HEADERS: The headers of the outgoing request is stripped of any headers that are not present in the 
         mock definition headers and then sorted and normalized before comparison.
METHOD:  The method of the outgoing request is compared as is.
------------------------------------------------------------------------------------------------------------------

Request:

  ┌────────────────────┬──────────────────────────────────────────────────────────────────────┐
  │  Outbound Request  │                                Value                                 │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Url                │ http://example.org/path?query[parameter]=1&query[parameter2]=2       │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Method             │ get                                                                  │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Body               │ maybe_gql=query+{+user+{+id+}+}&id=1&parameter=1                     │
  ├────────────────────┼──────────────────────────────────────────────────────────────────────┤
  │ Headers            │ header1=value&header2=value2                                         │
  └────────────────────┴──────────────────────────────────────────────────────────────────────┘

Mock definitions installed:


  ┌───────────────────────────────────────────────┬───────────────────────────────────────────────┐
  │              Colors Explanation               │               Diff Explanation                │
  ├───────────────────────────────────────────────┼───────────────────────────────────────────────┤
  │                                               │                                               │
  │ ● Exact match                                 │ Mock value matches the request value          │
  │ ● Partial match (When URL and METHOD match)   │ Present in mock but not in request            │
  │ ● Not matching                                │ Present in request but not in mock            │
  │                                               │                                               │
  └───────────────────────────────────────────────┴───────────────────────────────────────────────┘

NO MOCKS INSTALLED


[Mappersmith Test] There are no mocks installed, please refer to the documentation: https://github.com/tulios/mappersmith?tab=readme-ov-file#-testing-mappersmith  
"
`;
